frontend/catalyst/jax_tracer.py

```diff
@@ -689,45 +694,137 @@ def trace_quantum_function(
+            # out_tree_promise holds the output pytree structure for the
+            # original quantum function being trace. We assume that
+            # all transforms will preserve this type as the output.
+            #
+            # But this might not be strictly necessary.
             wffa, in_avals, out_tree_promise = deduce_avals(f, args, kwargs)
             in_classical_tracers = _input_type_to_tracers(trace.new_arg, in_avals)
             with QueuingManager.stop_recording(), quantum_tape:
                 # Quantum tape transformations happen at the end of tracing
                 ans = wffa.call_wrapped(*in_classical_tracers)
-            out_classical_tracers_or_measurements = [
-                (trace.full_raise(t) if isinstance(t, DynamicJaxprTracer) else t) for t in ans
-            ]
 
+            out_classical_tracers = []
+            out_measurements = []
+            out_classical_tracers_or_measurements = []
+            for ret_val in ans:
+                if isinstance(ret_val, DynamicJaxprTracer):
+                    ret_val_variable = trace.full_raise(ret_val)
+                    out_classical_tracers.append(ret_val_variable)
+                else:
+                    ret_val_variable = ret_val
+                    out_measurements.append(ret_val_variable)
+                out_classical_tracers_or_measurements.append(ret_val_variable)
+
+            params = quantum_tape.get_parameters(trainable_only=False)
+            quantum_tape.trainable_params = qml.math.get_trainable_indices(params)
+
+            if qnode and qnode.gradient_fn and isinstance(qnode.gradient_fn, qml.gradients.gradient_transform):
+                quantum_tape = qnode.gradient_fn.expand_fn(quantum_tape)
+
+            # Can transforms be applied?
+            # Since transforms are a PL feature and PL does not support the same things as
+            # Catalyst, transforms may have invariants that rely on PL invariants.
+            # For example:
+            #   * mid-circuit measurements (for batch-transforms)
+            #   * that the output will be only a sequence of `MeasurementProcess`es.
+            is_measurement = lambda maybe: isinstance(maybe, MeasurementProcess)
+            is_out_measurements = map(is_measurement, out_measurements)
+            is_all_out_measurements = all(is_out_measurements) and not out_classical_tracers
+            is_out_measurement_sequence = is_all_out_measurements and isinstance(out_tree_promise().unflatten(out_measurements), Sequence)
+            is_out_measurement_sequence_one_element = is_out_measurement_sequence and len(out_tree_promise().unflatten(out_measurements)) == 1
+            is_out_single_measurement = is_all_out_measurements and isinstance(out_tree_promise().unflatten(out_measurements), MeasurementProcess)
+            is_valid_output = is_out_measurement_sequence or is_out_single_measurement
+            # TODO: check if there were mid circuit measurements in the original tape.
```

Comment:
Is it something that you want to add in this PR or later?

Reply:
I added a check now and some tests for the validity of the transform: https://github.com/PennyLaneAI/catalyst/pull/280/commits/3812373f2b9bb0ca5ca8c3f95df8893207b0271b https://github.com/PennyLaneAI/catalyst/pull/280/commits/7a15d3f46ae9807309f681b17477f515bdea22b0